src/middlewared/middlewared/plugins/filesystem_/acl_linux.py

```diff
@@ -12,86 +14,112 @@ class FilesystemService(Service, ACLBase):
 
     @private
     def acltool(self, path, action, uid, gid, options):
+
+        flags = "-r"
+        flags += "x" if options.get('traverse') else ""
+        flags += "C" if options.get('do_chmod') else ""
+        flags += "P" if options.get('posixacl') else ""
+
         acltool = subprocess.run([
-            '/usr/bin/acltool',
+            '/usr/bin/nfs4xdr_winacl',
             '-a', action,
             '-O', str(uid), '-G', str(gid),
-            '-rx' if options['traverse'] else '-r',
+            flags,
             '-c', path,
             '-p', path], check=False, capture_output=True
         )
         if acltool.returncode != 0:
             raise CallError(f"acltool [{action}] on path {path} failed with error: [{acltool.stderr.decode().strip()}]")
 
-    def _common_perm_path_validate(self, path):
-        if not os.path.exists(path):
-            raise CallError(f"Path not found: {path}",
-                            errno.ENOENT)
+    def _common_perm_path_validate(self, schema, path, recursive, verrors):
+        p = Path(path)
+        if not p.is_absolute():
+            verrors.add(f'{schema}.path', 'Must be an absolute path.')
+            return
+
+        if p.is_file() and recursive:
+            verrors.add(f'{schema}.path', 'Recursive operations on a file are invalid.')
+            return
 
         if not os.path.realpath(path).startswith('/mnt/'):
-            raise CallError(f"Changing permissions on paths outside of /mnt is not permitted: {path}",
-                            errno.EPERM)
+            verrors.add(
+                f'{schema}.path',
+                "Changes to permissions on paths that are not beneath "
+                f"the directory /mnt are not permitted: {path}"
+            )
 
-        if os.path.realpath(path) in [x['path'] for x in self.middleware.call_sync('pool.query')]:
-            raise CallError(f"Changing permissions of root level dataset is not permitted: {path}",
-                            errno.EPERM)
+        elif len(p.resolve().parents) == 2:
+            verrors.add(
+                f'{schema}.path',
+                f'The specified path is a ZFS pool mountpoint "({path})" '
```

Comment:
Maybe change wording to something like you have right above `Changes to permissions for the ZFS pool mountpoint is not permitted` or something.